### PR TITLE
[JENKINS-36570] Implement pipeline compatibility

### PR DESCRIPTION
**Pull request by vitalii-vitrenko**
[[JENKINS-36570]](https://issues.jenkins-ci.org/browse/JENKINS-36570)

### CODE REVIEW

File: src/main/java/jenkins/advancedqueue/PriorityConfiguration.java [hunk 1]
```diff
@@ -207,6 +208,9 @@ public PriorityConfigurationCallback getPriority(Queue.Item item, PriorityConfig
 	}
 
 	private PriorityConfigurationCallback getPriorityInternal(Queue.Item item, PriorityConfigurationCallback priorityCallback) {
+		if (item.task instanceof ExecutorStepExecution.PlaceholderTask) {
```
Comment (oleg-nenashev) [inline src/main/java/jenkins/advancedqueue/PriorityConfiguration.java]: It will cause ClassNotFound exceptions if the Pipeline plugin is not installed

**Comment by Vestail** [comment]
Solved ClassNotFoundException issue just like it was done for matrix-project.

File: src/main/java/jenkins/advancedqueue/PriorityConfigurationPlaceholderTaskHelper.java [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+import hudson.model.Queue;
+import jenkins.advancedqueue.sorter.ItemInfo;
+import jenkins.advancedqueue.sorter.QueueItemCache;
+import jenkins.model.Jenkins;
+import org.jenkinsci.plugins.workflow.support.steps.ExecutorStepExecution;
+
+class PriorityConfigurationPlaceholderTaskHelper {
+
+    boolean isPlaceholderTask(Queue.Task task) {
+        return task instanceof  ExecutorStepExecution.PlaceholderTask;
```
Comment (oleg-nenashev) [inline src/main/java/jenkins/advancedqueue/PriorityConfigurationPlaceholderTaskHelper.java]: I would add a isPlaceholderTaskUsed() check before invoking instanceOf. Or just check class name by a string comparison. Both options should prevent exception here in the case of API misusage

File: src/main/java/jenkins/advancedqueue/PriorityConfiguration.java [hunk 3]
```diff
@@ -110,6 +112,10 @@ public int compare(JobGroup.PriorityStrategyHolder o1, JobGroup.PriorityStrategy
 		} else {
 			priorityConfigurationMatrixHelper = new PriorityConfigurationMatrixHelper();
 		}
+
+		if (PriorityConfigurationPlaceholderTaskHelper.isPlaceholderTaskUsed()) {
```
Comment (oleg-nenashev) [inline src/main/java/jenkins/advancedqueue/PriorityConfiguration.java]: There is no plugin existence check, hence it does not solve the exception issue

Comment (Vestail) [inline src/main/java/jenkins/advancedqueue/PriorityConfiguration.java]: `PriorityConfigurationPlaceholderTaskHelper.isPlaceholderTaskUsed()`  return  false if `Jenkins.getInstance().getPlugin("workflow-durable-task-step") == null` it is actually a plugin existence check.

**Comment by Vestail** [comment]
Did some refactoring.

File: src/main/java/jenkins/advancedqueue/PriorityConfiguration.java (hunk 3 above)
Comment (oleg-nenashev) [inline src/main/java/jenkins/advancedqueue/PriorityConfiguration.java]: right

File: src/main/java/jenkins/advancedqueue/PriorityConfigurationPlaceholderTaskHelper.java [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+import hudson.model.Job;
+import hudson.model.Queue;
+import jenkins.advancedqueue.sorter.ItemInfo;
+import jenkins.advancedqueue.sorter.QueueItemCache;
+import jenkins.model.Jenkins;
+import org.jenkinsci.plugins.workflow.support.steps.ExecutorStepExecution;
+
+class PriorityConfigurationPlaceholderTaskHelper {
+
+    private static boolean placeholderTaskUsed = Jenkins.getInstance().getPlugin("workflow-durable-task-step") != null;
```
Comment (oleg-nenashev) [inline src/main/java/jenkins/advancedqueue/PriorityConfigurationPlaceholderTaskHelper.java]: 🐛 Is not going to work if you dynamically install Pipeline, restart will be required

**Comment by timothyclarke** [comment]
Is there anything still outstanding on this one ?

File: src/main/java/jenkins/advancedqueue/PriorityConfigurationPlaceholderTaskHelper.java [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+    PriorityConfigurationCallback getPriority(ExecutorStepExecution.PlaceholderTask task, PriorityConfigurationCallback priorityCallback) {
+        Job<?, ?> job = (Job<?, ?>) task.getOwnerTask();
+        ItemInfo itemInfo = QueueItemCache.get().getItem(job.getName());
+        itemInfo.getPriority();
+        priorityCallback.setPrioritySelection(itemInfo.getPriority());
+        return priorityCallback;
+    }
+
+    static boolean isPlaceholderTaskUsed() {
+        return Jenkins.getInstance().getPlugin("workflow-durable-task-step") != null;
```
Comment (oleg-nenashev) [inline src/main/java/jenkins/advancedqueue/PriorityConfigurationPlaceholderTaskHelper.java]: It also makes sense to ensure the plugin is not disabled

Comment (Vestail) [inline src/main/java/jenkins/advancedqueue/PriorityConfigurationPlaceholderTaskHelper.java]: For what? `PriorityConfigurationPlaceholderTaskHelper.isPlaceholderTask()` will always return `false` in this case.

Comment (oleg-nenashev) [inline src/main/java/jenkins/advancedqueue/PriorityConfigurationPlaceholderTaskHelper.java]: After checking the code I'm not 100% sure about it. The plugin may be registered, but not active yet. In such case classloader won't be able to find the class. This method will return true, but then isPlaceholderTask() will fail with ClassNotFoundException. Need to reproduce it, but it's what I've seen on old core versions at least

**Comment by samskiter** [comment]
Is there any doc for how to use this in a pipeline? @oleg-nenashev @vitalii-vitrenko

**Comment by samskiter** [comment]
(we would like to prioritise later *stages* within our pipeline higher than earlier ones - will this plugin allow this?)

**Comment by oleg-nenashev** [comment]
No, it's not possible. Documentation PRs will be appreciated

**Comment by samskiter** [comment]
I have no idea what this does then. So I'm afraid I cannot document it!

**Comment by oleg-nenashev** [comment]
Sounds fair. It just sets the priority on the Pipeline's run level. Pipeline submits subtasks to acquire agents, and this PR just makes these subtasks to inherit priority of the Run. Nothing more from what I see

**Comment by samskiter** [comment]
could this plugin be easily modified to set the priority of these so-called subtasks to achieve the goal I mentioned?

**Comment by oleg-nenashev** [comment]
No, in general it cannot. This plugin allows managing priorities on the `node()` level for Pipeline, but there is no API in core/Pipeline which would allow to manage `stage()` priorities. Just because stages happen within the same task. If each of your stages runs on a separate node, then it's possible to update this plugin (cc @svanoort @abayer @rsandell ).

Anyway, I suggest creating a feature request in Jenkins JIRA. Discussion in a closed PR will just get lost at some point

**Comment by samskiter** [comment]
Yes, we'd like to run later stages sooner.

I found a couple of others after the same issue on SO here: https://stackoverflow.com/questions/44270982/how-can-i-prioritize-later-steps-when-running-jenkins-pipeline-scripts

To respond to your points:

1.  Add build agents. In the era of on-demand cloud hardware, this is pretty trivial.

If your business requirement is on-premise this is less trivial
We also require macos machines which cannot be virtualised, again making it less trivial

2.  Use milestone steps to abort redundant builds of the same job

Redundancy isn't the issue - we already have that solved. The SO post above outlines the issue better than I can state

3.  Otherwise throttle/limit some builds to free up build agents

Throttling is what are are doing for now, but it gives us poor utilisation.

**Pull request merged**